ephios/core/views/auth.py

```diff
@@ -0,0 +1,167 @@
+from requests import PreparedRequest, RequestException
+from requests_oauthlib import OAuth2Session
+
+from ephios.core.forms.users import OIDCDiscoveryForm
+from ephios.core.models.users import IdentityProvider
+
+
+class OIDCInitiateView(RedirectView):
+    def get_redirect_url(self, *args, **kwargs):
+        provider = get_object_or_404(IdentityProvider, id=self.kwargs["provider"])
+
+        # we are using the OAuth2 tooling here to generate the authorization URL
+        oauth_client = WebApplicationClient(client_id=provider.client_id)
+        oauth = OAuth2Session(
+            client=oauth_client,
+            redirect_uri=urljoin(settings.GET_SITE_URL(), reverse("core:oidc_callback")),
+            scope=provider.scopes,
+        )
+
+        authorization_url, state = oauth.authorization_url(provider.authorization_endpoint)
+        self.request.session["oidc_state"] = state
+        self.request.session["oidc_provider"] = provider.id
+        self.request.session["oidc_login_next"] = self.request.GET.get("next", None)
+        return authorization_url
+
+
+class OIDCCallbackView(RedirectView):
+    def failure_url(self):
+        messages.error(self.request, _("Authentication failed."))
+        return settings.LOGIN_URL
+
+    def get_redirect_url(self, *args, **kwargs):
+        if "error" in self.request.GET:
+            return self.failure_url()
+        if "state" in self.request.GET and "code" in self.request.GET:
+            if (
+                "oidc_state" not in self.request.session
+                or self.request.session.pop("oidc_state") != self.request.GET["state"]
+            ):
+                return self.failure_url()
+
+            user = auth.authenticate(self.request)
+
+            if user and user.is_active:
+                request_user = getattr(self.request, "user", None)
+                if not request_user or not request_user.is_authenticated or request_user != user:
+                    auth.login(self.request, user)
+                return self.request.session.get("oidc_login_next") or "/"
+        return self.failure_url()
+
+
+class OIDCLogoutView(RedirectView):
+    def get_redirect_url(self, *args, **kwargs):
+        logout_url = reverse("login")
+        if "oidc_provider" in self.request.session:
+            providers = IdentityProvider.objects.filter(
+                id=self.request.session.get("oidc_provider")
+            )
+            if providers.exists() and (provider := providers.first()).end_session_endpoint:
+                req = PreparedRequest()
+                req.prepare_url(
+                    provider.end_session_endpoint,
+                    {"post_logout_redirect_uri": settings.GET_SITE_URL()},
+                )
+                logout_url = req.url
+        auth.logout(self.request)
+        messages.info(self.request, _("Logged out successfully."))
+        return logout_url
+
+
+class IdentityProviderCreateView(SuccessMessageMixin, CreateView):
```

Comment:
![image](https://github.com/ephios-dev/ephios/assets/5970076/f235ed75-dcd4-4ec7-8fa9-6a503cdfbe12)
Template seems not right to me. missing menu entries

Reply:
also I got logged out when creating the provider?

Reply:
jep, noticed that yesterday when going to sleep ^^

Reply:
I cannot reproduce being logged out after creating the provider?

Reply:
> I cannot reproduce being logged out after creating the provider?

I probably logged out in another tab... sorry!